### DISCUSSION TITLE
Disk affinity/anti-affinity proposal discussion

### DISCUSSION THREAD

**Discussion by mishoyama** [dell/csi-baremetal#715] [General]
Link to the proposal - https://github.com/dell/csi-baremetal/blob/master/docs/proposals/disk-affinity.md

@Aber4Nod please see my comments:

1. https://github.com/dell/csi-baremetal/blob/master/docs/proposals/disk-affinity.md#abstract doesn't provide description which applications might need this feature and what is the use case
2. https://github.com/dell/csi-baremetal/blob/master/docs/proposals/disk-affinity.md#proposal is really hard to understand and read. A lot of things a put together and mixed up. For example:
```
affinity.volumes.csi-baremetal.dell.com/types: <pod-bound-required/pod-bound-preferred/volume-bound-required/volume-bound-preferred/dedicated-required/dedicated-preferred
```
If you split into two parts it's becoming more clear for understanding:
```
affinity.volumes.csi-baremetal.dell.com/type: pod-bound/volume-bound/dedicated
affinity.volumes.csi-baremetal.dell.com/mode: required/preferred
```
The same for `antiaffinity` annotations
3. If you want to use annotations you need to clearly say where to put them and don't postpone this till the implementation section
4. Proposal should start with use cases which you plan to address by this feature
5. After reading proposal I still don't understand which API should I use two cover the following use case:
- Node with 2 SSDs and _X_ HDDs
- Application 1 with 3 PVCs wants to have dedicated SSD using SSD LVG SC
- _X_ other applications to use generic ephemeral volumes on remaining SSD
5. Other questions:
- What is the use case for `antiaffinity.volumes.csi-baremetal.dell.com/labels`? Do we want latency sensitive applications to not share single disk? But other application can use it?
- How will this work with ephemeral volumes?
- Can we use SC API which is always configurable?

**Comment by safronovD** [comment]
Could we separate logic of the pod groups creating and selecting affinity policies for groups?
For example, we have options to create patterns based on pod names:
```
group1: [*zookeeper*, *monitoring*]
group2: [*]
```
or to labels pods with group names:
```
zookeper-0
   labels=[affinity.csi-baremetal.dell.com/selector=group1]
zookeper-1
   labels=[affinity.csi-baremetal.dell.com/selector=group1]
other-pod-0
   labels=[affinity.csi-baremetal.dell.com/selector=group2]
```
And after this operation user should create ConfigMap or CR with description of affinity policies:
```
dedicated_groups: [group1]
affinity:
    preferred: [group2, group3]
antiaffinity:
    required: [group2, group4]
```
Also we can implement option for user to choose LVG manually:
```
affinity:
    preferred: 
        pod_selector: [group1, group2]
        lvg_name: lvg1 # If it is not set, CSI select LVG automatically from free disks
```
